python/ql/lib/semmle/python/dataflow/new/internal/TypeTracker.qll

```diff
@@ -69,7 +69,10 @@ private module Cached {
    */
   cached
   predicate stepCall(TypeTrackingNode nodeFrom, TypeTrackingNode nodeTo, StepSummary summary) {
-    exists(Node mid | nodeFrom.flowsTo(mid) and smallstepCall(mid, nodeTo, summary))
+    exists(Node mid | nodeFrom.flowsTo(mid) and smallstepCall(mid, nodeTo, summary)) and
+    // performance tweak: if the source is a parameter, then flow will always originate
+    // from a call, so we can cut away returning flow
+    if nodeFrom instanceof ParameterNode then summary != ReturnStep() else any()
```

Comment:
I think this is too strong. What if I want to start a type tracker in a specific argument of a specific function, say

```python
def handle(request):
    # ... do various stuff
    return request
```
and then track the return value to a sink in a function that calls `handle`? By my reading, this would be impossible since we've disallowed return steps from `request`. Sure, `request` must come from a call, but it might not be desirable to start the type tracking there (since then it won't be able to exit `handle` again).

Or am I missing something here? :thinking:

Reply:
Indeed, type tracking starting at parameters and returning out will no longer work. So if that is something we need to support, I'll have to think of something else.

Reply:
Hmm... So, if we had unified return nodes (as `TypeTrackingNode`s), then there would be an obvious workaround: start the type tracking at the return node instead (or perhaps the union of the two, to also track _inside_ the handler). 
I guess I don't really know what problem you're trying to fix, so I'm not sure how helpful I can be. :neutral_face:

Reply:
> So, if we had unified return nodes. 

In JavaScript with have a `DataFlow::FunctionReturnNode` (and `DataFlow::ExceptionalFunctionReturnNode`).  
It sounds similar to what you are talking about.